docs/fsharp-cheatsheet.md

```diff
@@ -540,17 +537,106 @@ Another way of implementing interfaces is to use *object expressions*.
+        | DivisibleBy 5 -> "Buzz"
         | i -> string i
 
 *Partial active patterns* share the syntax of parameterized patterns but their active recognizers accept only one argument.
 
+<a name="CodeOrganization"></a>Code Organization
+---------
+
+### Modules
+Modules are key building blocks for grouping related concepts; they can contain `types`, `let` bindings, and even other `modules`.
+Modules can be referenced using dot notation or exposed with the `open` keyword. Illustrative-only example:
+
+    module Game =
+        let mutable basePoints = 1
+        type Player = { id: int; score: int }
+        let playerScored player = { player with score = player.score + basePoints }
+
+    let player: Game.Player = { id = 1; score = 0 }
+    let player' = Game.playerScored player  // score = 1
+
+    open Game
+    basePoints <- 2
+    let player'' = playerScored player'  // player''.score = 3
+
+If you have only one module in your file, the `module` can be specified at the top of the file.
+
+    module Functions  // notice there is no '=' when at the top of a file
+
+    let addToFive num = 5 + num
+    let subtractFive num = num - 5
+
+### Namespaces
+Namespaces are simply dotted strings that prefix names. They are placed at the top of the file.
+An inner namespace can be specified in the same file.
+
+    namespace MyNamespace
+
+    namespace MyNamespace.InnerSpace
+
+They can also be part of a module name.
+
+    module MyNamespace.InnerSpace.MyModule
+
+### Open and AutoOpen
+
+The `open` keyword can be used with `module`, `namespace`, and `type`.
+
+    open System.Diagnostics  // open namespace
+    let stopwatch = Stopwatch.StartNew()
+    ---
+    module MyModule =
+        type DU1 = A | B | C
+        type DU2 = D | E | F
+
+    open type MyModule.DU1
+    let du1 = A
+    let duNotDefined = D  // 'D' not defined
+    open MyModule
+    let du2 = D
+    ---
+    [<AutoOpen>]
+    module MyModule =
+        type DU = A | B | C
+
+    let du = A
+
+### Accessibility Modifiers
+
+F# supports `public`, `private` and `internal`.  It can be applied to `module`, `let`, `member`, and `type`.
+
+    module private MyModule = ...
+    ---
+    let private x = ...
+    ---
+    member private self.x = ...
+    ---
+    type private Person = ...
+
+### Recursive Reference
+
+F#'s dependency resolution is based on file and code order. This requirement encourages developers to think about the design of their programs and dependencies upfront,
+which results in cleaner, maintainable code, but in rare cases you may need to loosen those rules.
+To do this we have `rec` for `module` and `namespace`s; and `and` for `type`s and `let` functions.
+
+    module rec MyNamespace.MonkeyDomain
+
+    exception DoNotSqueezeBananaException of Banana // `Banana` has not been defined yet, and would fail without `rec`
+
+    type Banana = { Type: string; IsRipe: bool }
+        member self.Squeeze() = raise (DoNotSqueezeBananaException self)
```

Comment:
better to bring `_` vs `this` vs `self` into play elsewhere
```suggestion
        member _.Squeeze() = raise (DoNotSqueezeBananaException self)
```

Reply:
We need the `self` to add to `DoNotSqueezeBananaException`. However, after looking at it again, I realize *I* have a bug because I need a `with` declaration before `member`. :-D

Reply:
ah..., missed that
but... better to format without a `with` by doing
```
type X =
    { .... }
    member self_.Squeeze()
```
(F# Style guide talks about this a bit IIRC)